### ISSUE TITLE
[PATCH] refactor macro to avoid compiler warning in regcomp.c

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11980]
Migrated from [rt.perl.org#111468](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=111468) (status was 'resolved')

Searchable as RT111468$

**Comment by p5pRT** [comment]
### From rmbarker.cpan@btinternet.com

#### Created by rmbarker.cpan@btinternet.com



The compiler warning in regcomp\.c \(as noted in recent comment\)
can be avoided by refactoring DO\_POSIX\_LATIN1\_ONLY\_KNOWN
to separate the case where sourcelist==l1\_sourcelist

Patch attached\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.15.8:

Configured by robin at Tue Feb 28 21:50:27 GMT 2012.

Summary of my perl5 (revision 5 version 15 subversion 8) configuration:
  Local Commit: 789f7999e0b8bb756ff1596fef9ac3bbeee43892
  Ancestor: 4d97b3cc28cd777c5607474f9a98679b04f88a55
  Platform:
    osname=linux, osvers=2.6.38-13-generic, archname=i686-linux-64int
    uname='linux spade-ubuntu 2.6.38-13-generic #55-ubuntu smp tue jan
24 14:27:59 utc 2012 i686 i686 i386 gnulinux '
    config_args='-ders -Dcc=gcc -Dusedevel -Doptimize=-O2 -DDEBUGGING
-Duse64bitint -Dman3dir=none -Dcf_email=rmbarker.cpan@btinternet.com'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-DDEBUGGING -fno-strict-aliasing -pipe
-fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE
-D_FILE_OFFSET_BITS=64',
    optimize='-O2 -g',
    cppflags='-DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector
-I/usr/local/include'
    ccversion='', gccversion='4.6.1', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long long', ivsize=8, nvtype='double', nvsize=8,
Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -fstack-protector -L/usr/local/lib'

libpth=/usr/local/lib /lib /usr/lib /usr/lib/i386-linux-gnu /usr/lib64
    libs=-lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.13'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -g -L/usr/local/lib
-fstack-protector'

Locally applied patches:
    


@INC for perl 5.15.8:
    lib
    /usr/local/lib/perl5/site_perl/5.15.8/i686-linux-64int
    /usr/local/lib/perl5/site_perl/5.15.8
    /usr/local/lib/perl5/5.15.8/i686-linux-64int
    /usr/local/lib/perl5/5.15.8
    /usr/local/lib/perl5/site_perl
    .


Environment for perl 5.15.8:
    HOME=/home/robin
    LANG=en_GB.UTF-8
    LANGUAGE=en_GB:en
    LD_LIBRARY_PATH=/usr/lib:/usr/local/lib
    LOGDIR (unset)

PATH=/home/robin/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games
    PERL_BADLANG (unset)
    SHELL=/bin/tcsh


```
</details>

**Comment by p5pRT** [comment]
### From rmbarker.cpan@btinternet.com

<details><summary>refactor-macro.patch</summary>

```diff
Subject: refactor macro to avoid compiler warning in regcomp.c

The compiler warning in regcomp.c (as noted in recent comment)
can be avoided by refactoring DO_POSIX_LATIN1_ONLY_KNOWN
to separate the case where sourcelist==l1_sourcelist

---
diff --git a/regcomp.c b/regcomp.c
index 7a3433a..b848af3 100644
--- a/regcomp.c
+++ b/regcomp.c
@@ -10121,6 +10121,12 @@ S_checkposixcc(pTHX_ RExC_state_t *pRExC_state)
  * time */
 #define DO_POSIX_LATIN1_ONLY_KNOWN(node, class, destlist, sourcelist,      \
                               l1_sourcelist, Xpropertyname, run_time_list) \
+        DO_POSIX_ONLY_KNOWN(node, class, destlist,                         \
+                ((AT_LEAST_ASCII_RESTRICTED) ? sourcelist : l1_sourcelist),\
+                Xpropertyname, run_time_list) 
+
+#define DO_POSIX_ONLY_KNOWN(node, class, destlist, sourcelist,             \
+                Xpropertyname, run_time_list)                              \
     /* If not /a matching, there are going to be code points we will have  \
      * to defer to runtime to look-up */                                   \
     if (! AT_LEAST_ASCII_RESTRICTED) {                                     \
@@ -10130,11 +10136,7 @@ S_checkposixcc(pTHX_ RExC_state_t *pRExC_state)
         ANYOF_CLASS_SET(node, class);                                      \
     }                                                                      \
     else {                                                                 \
-        _invlist_union(destlist,                                           \
-                       (AT_LEAST_ASCII_RESTRICTED)                         \
-                           ? sourcelist                                    \
-                           : l1_sourcelist,                                \
-                       &destlist);                                         \
+        _invlist_union(destlist, sourcelist, &destlist);                   \
     }
 
 /* Like DO_POSIX_LATIN1_ONLY_KNOWN, but for the complement.  A combination of
@@ -10853,10 +10855,8 @@ parseit:
                                             PL_PosixCntrl, PL_XPosixCntrl);
 		    break;
 		case ANYOF_DIGIT:
-		    /* Ignore the compiler warning for this macro, planned to
-		     * be eliminated later */
-		    DO_POSIX_LATIN1_ONLY_KNOWN(ret, namedclass, properties,
-                        PL_PosixDigit, PL_PosixDigit, "XPosixDigit", listsv);
+		    DO_POSIX_ONLY_KNOWN(ret, namedclass, properties,
+                        PL_PosixDigit, "XPosixDigit", listsv);
 		    break;
 		case ANYOF_NDIGIT:
 		    DO_N_POSIX_LATIN1_ONLY_KNOWN(ret, namedclass, properties,

```
</details>

**Comment by p5pRT** [comment]
### From @khwilliamson

On 03/01/2012 08&#8203;:20 AM\, Robin Barker \(via RT\) wrote&#8203;:
> \# New Ticket Created by  Robin Barker
> \# Please include the string&#8203;:  \[perl \#111468\]
> \# in the subject line of all future correspondence about this issue\.
> \#\<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=111468>
>
>
> This is a bug report for perl from rmbarker\.cpan@&#8203;btinternet\.com\,
> generated with the help of perlbug 1\.39 running under perl 5\.15\.8\.
>
>
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
>
> The compiler warning in regcomp\.c \(as noted in recent comment\)
> can be avoided by refactoring DO\_POSIX\_LATIN1\_ONLY\_KNOWN
> to separate the case where sourcelist==l1\_sourcelist
>
> Patch attached\.
>

We had decided in irc to leave the message in for 5\.16\, as there are 
similar messages in the build\, and this is planned to be fixed in 5\.17\.

I have no objections to this patch\, and will apply it if we deem this as 
something to fix in 5\.16\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

Fixed in 4f6289a38\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
